services/src/main/java/org/keycloak/vault/AbstractVaultProviderFactory.java

```diff
@@ -0,0 +1,174 @@
+ * for the configuration of key resolvers.
+ * <p/>
+ * It implements the {@link #init(Config.Scope)} method, where is looks for the {@code keyResolvers} property. The value is
+ * a comma-separated list of key resolver names. It then verifies if the resolver names match one of the available key resolver
+ * implementations and then creates a list of {@link VaultKeyResolver} instances that subclasses can pass to {@link VaultProvider}
+ * instances on {@link #create(KeycloakSession)}.
+ * <p/>
+ * The list of currently available resolvers follows:
+ * <ul>
+ *     <li>{@code KEY_ONLY}: only the key name is used as is, realm is ignored;</li>
+ *     <li>{@code REALM_UNDERSCORE_KEY}: realm and key are combined using an underscore ({@code '_'}) character. Any occurrences of
+ *     underscore in both the realm and key are escaped by an additional underscore character;</li>
+ *     <li>{@code REALM_FILESEPARATOR_KEY}: realm and key are combined using the platform file separator character. It might not be
+ *     suitable for every vault provider but it enables the grouping of secrets using a directory structure;</li>
+ *     <li>{@code FACTORY_PROVIDED}: the format of the constructed key is determined by the factory's {@link #getFactoryResolver()}
+ *     implementation. it allows for the customization of the final key format by extending the factory and overriding the
+ *     {@link #getFactoryResolver()} method.</li>
+ * </ul>
+ * <p/>
+ * <b><i>Note</i></b>: When extending the standard factories to use the {@code FACTORY_PROVIDED} resolver, it is important to also
+ * override the {@link #getId()} method so that the custom factory has its own id and as such can be configured in the keycloak
+ * server.
+ * <p/>
+ * If no resolver is explicitly configured for the factory, it defaults to using the {@code REALM_UNDERSCORE_KEY} resolver.
+ * When one or more resolvers are explicitly configured, this factory iterates through them in order and for each one attempts
+ * to obtain the respective {@link VaultKeyResolver} implementation. If it fails (for example, the name doesn't match one of
+ * the existing resolvers), it logs a message and ignores the resolver. If it fails to load all configured resolvers, it
+ * throws a {@link VaultConfigurationException}.
+ * <p/>
+ * Concrete implementations must also make sure to call the {@code super.init(config)} in their own {@link #init(Config.Scope)}
+ * implementations so tha the processing of the key resolvers is performed correctly.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public abstract class AbstractVaultProviderFactory implements VaultProviderFactory {
+
+    private static final Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
+
+    protected static final String KEY_RESOLVERS = "keyResolvers";
+
+    protected List<VaultKeyResolver> keyResolvers = new ArrayList<>();
```

Comment:
AFAIK This line initialize List with size for 10 items in memory, however, in the majority of cases, there will be only one/two items in the end. We can initialize it for 1/2 items or initialize it lazily based on the size of resolvers provided in config or to 1 in case KEY_RESOLVERS configuration is missing. @sguilhen WDYT? Is this kind of optimization relevant for us?

Reply:
Since this list is only used for adding and full iteration (no indexed access), `LinkedList` is best option.

Reply:
Agreed, changed to LinkedList